File: ios/new/HybridViewModelImageProperty.swift
import RiveRuntime
import NitroModules

class HybridViewModelImageProperty: HybridViewModelImagePropertySpec {
  private var instance: ViewModelInstance?
  private var prop: ImageProperty?
  private var worker: Worker?

  /// Bumped by every set(). Decoding is async, so a slow decode can finish after a later
  /// set() has already applied — the generation it captured lets it detect that and bail.
  @MainActor private var generation: UInt64 = 0

  init(instance: ViewModelInstance, path: String, worker: Worker) {
    self.instance = instance
    self.prop = ImageProperty(path: path)
    self.worker = worker
    super.init()
  }

  override init() {
    super.init()
  }

  func set(image: (any HybridRiveImageSpec)?) throws {
    guard let instance = instance, let prop = prop, let worker = worker else {
      throw RuntimeError.error(withMessage: "ImageProperty not properly initialized")
    }
    guard let image = image else {
      Task { @MainActor in
        self.generation &+= 1
        instance.setValue(of: prop, to: nil)
      }
      return
    }
    guard let hybridImage = image as? HybridRiveImage else {
      throw RuntimeError.error(withMessage: "Invalid image type - expected HybridRiveImage")
    }

    Task { @MainActor in
      self.generation &+= 1
      let generation = self.generation
      do {
        let experimentalImage = try await worker.decodeImage(from: hybridImage.rawData)
        guard generation == self.generation else { return }
        instance.setValue(of: prop, to: experimentalImage)
      } catch {
        RCTLogError("HybridViewModelImageProperty: Failed to decode/set image: \(error)")
      }
    }
  }

  func addListener(onChanged: @escaping () -> Void) throws -> () -> Void {
    // TODO: image property listener not yet available in concurrency API
    return {}
  }

  func removeListeners() throws {}

  func dispose() throws {}

  deinit {}
}

File: android/src/new/java/com/margelo/nitro/rive/HybridViewModelImageProperty.kt
package com.margelo.nitro.rive

import android.util.Log
import androidx.annotation.Keep
import app.rive.ImageAsset
import app.rive.ViewModelInstance
import app.rive.core.CommandQueue
import com.facebook.proguard.annotations.DoNotStrip
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicLong

@Keep
@DoNotStrip
class HybridViewModelImageProperty(
  private val instance: ViewModelInstance,
  private val path: String,
  private val riveWorker: CommandQueue
) : HybridViewModelImagePropertySpec(),
  BaseHybridViewModelProperty<Unit> by BaseHybridViewModelPropertyImpl() {
  companion object {
    private const val TAG = "HybridViewModelImageProperty"
  }

  private val imageScope = CoroutineScope(Dispatchers.Default)

  /**
   * Bumped by every set(). Decoding is async, so a slow decode can finish after a later set()
   * has already applied — the generation it captured lets it detect that and bail.
   */
  private val generation = AtomicLong(0)

  override fun set(image: HybridRiveImageSpec?) {
    if (image == null) {
      // rive-android's ViewModelInstance.setImage only accepts a non-null ImageAsset up to
      // 11.7.2. The nullable overload that clears the property landed upstream in
      // rive-app/rive-android#13261 and ships in the next release; until we bump the pin,
      // dropping the call is all we can do here. Clearing works on legacy and on iOS.
      Log.w(TAG, "Clearing image property '$path' is not supported by this rive-android version")
      return
    }
    val hybridImage = image as? HybridRiveImage ?: return
    val setGeneration = generation.incrementAndGet()
    imageScope.launch {
      try {
        val result = ImageAsset.fromBytes(riveWorker, hybridImage.rawData)
        if (result is app.rive.Result.Success) {
          if (generation.get() != setGeneration) return@launch
          instance.setImage(path, result.value)
        } else {
          Log.e(TAG, "Failed to decode image for path '$path'")
        }
      } catch (e: Exception) {
        Log.e(TAG, "Failed to set image for path '$path'", e)
      }
    }
  }

  override fun addListener(onChanged: () -> Unit): () -> Unit {
    // Image property listeners not supported in experimental API
    return {}
  }

  override fun removeListeners() {
    // no-op
  }
}
